canister/src/api/get_block_headers.rs

```diff
@@ -1,17 +1,130 @@
+    with_state, with_state_mut,
+};
+
+// The maximum number of block headers that are allowed to be included in a single
+// `GetBlockHeadersResponse`.
+const MAX_BLOCK_HEADERS_PER_RESPONSE: u32 = 100;
+
+// Various profiling stats for tracking the performance of `get_block_headers`.
+#[derive(Default, Debug)]
+struct Stats {
+    // The total number of instructions used to process the request.
+    ins_total: u64,
+
+    // The number of instructions used to build the block headers vec.
+    ins_build_block_headers_vec: u64,
+}
+
+fn verify_and_return_effective_range(
+    request: &GetBlockHeadersRequest,
+) -> Result<(u32, u32), GetBlockHeadersError> {
+    let chain_height = with_state(main_chain_height);
+
+    if request.start_height > chain_height {
+        return Err(GetBlockHeadersError::StartHeightDoesNotExist {
+            requested: request.start_height,
+            chain_height,
+        });
     }
-    Ok(())
+
+    let (effective_start_height, mut effective_end_height) =
+        if let Some(end_height) = request.end_height {
+            if end_height < request.start_height {
+                return Err(GetBlockHeadersError::StartHeightLagerThanEndHeight {
+                    start_height: request.start_height,
+                    end_height,
+                });
+            }
+
+            if end_height > chain_height {
+                return Err(GetBlockHeadersError::EndHeightDoesNotExist {
+                    requested: end_height,
+                    chain_height,
+                });
+            }
+            // If `end_height` is provided then it should be the
+            // end of effective height range.
+            (request.start_height, end_height)
+        } else {
+            // If `end_height`` is not provided then the end of effective
+            // range should be the last block of the chain.
+            (request.start_height, chain_height)
+        };
+
+    // Bound the length of block headers vec.
+    effective_end_height = std::cmp::min(
+        effective_end_height,
+        effective_start_height + MAX_BLOCK_HEADERS_PER_RESPONSE - 1,
+    );
+
+    Ok((effective_start_height, effective_end_height))
+}
+
+fn get_block_headers_internal(
+    request: &GetBlockHeadersRequest,
+) -> Result<(GetBlockHeadersResponse, Stats), GetBlockHeadersError> {
+    let (start_height, end_height) = verify_and_return_effective_range(request)?;
+
+    let mut stats: Stats = Stats::default();
+
+    // Build block headers vec.
+    let ins_start = performance_counter();
+
+    // Add requested block headers located in stable_blocks.
+    let mut vec_headers: Vec<Vec<u8>> = with_state(|s| {
+        let block_heights = &s.stable_block_headers.block_heights;
+        let block_headers = &s.stable_block_headers.block_headers;
+        block_heights
+            .range(start_height..=end_height)
+            .map(|(_, block_hash)| block_headers.get(&block_hash).unwrap().into())
+            .collect()
+    });
```

Comment:
I thought I wrote this yesterday. `block_heights` and `block_headers` in the `BlockHeaderStore` are supposed to be private. To that end, I think we can refactor this code to be a method in `BlockHeaderStore` instead. For example, it could have the following signature:

```
    pub fn get_block_headers_in_range(
        &self,
        heights: std::ops::RangeInclusive<Height>,
    ) -> impl Iterator<Item = BlockHeaderBlob> + '_ {
```

Reply:
Done.